File: packages/shared/src/protocol.ts
import { z } from 'zod';

// ---------- Primitives ----------

export const ClaimType = z.enum([
  'definition',
  'theorem',
  'lemma',
  'corollary',
  'proposition',
  'proof',
  'remark',
  'example',
  'claim',
  'conjecture',
  'exercise',
  'problem',
  'question',
]);
export type ClaimType = z.infer<typeof ClaimType>;

export const ClaimStatus = z.enum([
  'unverified',
  'translating',
  'verifying',
  'verified',
  'failed',
  'sorry',
]);
export type ClaimStatus = z.infer<typeof ClaimStatus>;

export const RootCauseCategory = z.enum([
  'unknownIdentifier',
  'typeMismatch',
  'tacticFailed',
  'timeout',
  'malformedJson',
  'other',
]);
export type RootCauseCategory = z.infer<typeof RootCauseCategory>;

export const MathlibWorthinessVerdict = z.enum([
  'mathlib_candidate',
  'project_local',
  'duplicate_likely',
  'needs_generalization',
  'not_worth_submitting',
  'unsure',
  'skipped',
]);
export type MathlibWorthinessVerdict = z.infer<typeof MathlibWorthinessVerdict>;

// ---------- Claim ----------

export const Claim = z.object({
  id: z.string(),
  type: ClaimType,
  number: z.string().optional(),
  label: z.string().optional(),
  startLine: z.number().int().nonnegative(),
  endLine: z.number().int().nonnegative(),
  statementLatex: z.string(),
  proofLatex: z.string().optional(),
  hashLatex: z.string(),
  hashNormalized: z.string(),
  status: ClaimStatus.default('unverified'),
  llmDependencyIds: z.array(z.string()).default([]),
});
export type Claim = z.infer<typeof Claim>;

// ---------- /verify_paper (orchestrated) ----------

export const VerifyPaperRequest = z.object({
  requestId: z.string().uuid(),
  targetClaimId: z.string(),
  claims: z.array(Claim),
  leanVersion: z.string(),
  mathlibVersion: z.string(),
});
export type VerifyPaperRequest = z.infer<typeof VerifyPaperRequest>;

// ---------- SSE events ----------

export const SseEvent = z.discriminatedUnion('type', [
  z.object({
    type: z.literal('orchestratorStarted'),
    requestId: z.string().uuid(),
    targetClaimId: z.string(),
    plan: z.array(
      z.object({ level: z.number().int().nonnegative(), claimIds: z.array(z.string()) }),
    ),
  }),
  z.object({
    type: z.literal('claimStatus'),
    requestId: z.string().uuid(),
    claimId: z.string(),
    status: ClaimStatus,
    detail: z.string().optional(),
  }),
  z.object({
    type: z.literal('claimVerified'),
    requestId: z.string().uuid(),
    claimId: z.string(),
    elapsedMs: z.number().int().nonnegative(),
    cacheHit: z.boolean(),
    leanCode: z.string().optional(),
  }),
  z.object({
    type: z.literal('mathlibWorthiness'),
    requestId: z.string().uuid(),
    claimId: z.string(),
    verdict: MathlibWorthinessVerdict,
    confidence: z.number().min(0).max(1),
    reason: z.string(),
    evidence: z.array(z.string()).default([]),
    suggestedLocation: z.string().optional(),
    suggestedName: z.string().optional(),
    reviewNotes: z.array(z.string()).default([]),
  }),
  z.object({
    type: z.literal('claimFailed'),
    requestId: z.string().uuid(),
    claimId: z.string(),
    rootCauseCategory: RootCauseCategory,
    deepestFailedClaimId: z.string(),
    explanation: z.string(),
    leanOutput: z.string().optional(),
    leanCode: z.string().optional(),
  }),
  z.object({
    type: z.literal('orchestratorFinished'),
    requestId: z.string().uuid(),
    overall: z.enum(['verified', 'failed', 'partial']),
  }),
  z.object({
    type: z.literal('claimDependencies'),
    requestId: z.string().uuid(),
    claimId: z.string(),
    llmDependencyIds: z.array(z.string()),
  }),
]);
export type SseEvent = z.infer<typeof SseEvent>;

// ---------- /status/{request_id} ----------

export const StatusResponse = z.object({
  requestId: z.string().uuid(),
  state: z.enum(['running', 'verified', 'failed', 'unknown']),
  events: z.array(SseEvent).default([]),
});
export type StatusResponse = z.infer<typeof StatusResponse>;

// ---------- Port RPC envelope (sidepanel <-> background) ----------

export const PortRequest = z.discriminatedUnion('type', [
  z.object({ type: z.literal('listClaims') }),
  z.object({ type: z.literal('verifyClaim'), claimId: z.string() }),
  z.object({ type: z.literal('cancelVerification'), requestId: z.string().uuid() }),
  z.object({ type: z.literal('getSettings') }),
  z.object({
    type: z.literal('setSettings'),
    settings: z.object({
      backendUrl: z.string().url(),
    }),
  }),
]);
export type PortRequest = z.infer<typeof PortRequest>;

export const PortMessage = z.discriminatedUnion('type', [
  z.object({ type: z.literal('claimsSnapshot'), claims: z.array(Claim) }),
  z.object({ type: z.literal('sse'), event: SseEvent }),
  z.object({
    type: z.literal('settings'),
    settings: z.object({
      backendUrl: z.string().url(),
    }),
  }),
  z.object({ type: z.literal('error'), message: z.string() }),
]);
export type PortMessage = z.infer<typeof PortMessage>;

export const PORT_NAME = 'lale-panel';

// ---------- Content <-> background ----------

export const ContentMessage = z.discriminatedUnion('type', [
  z.object({ type: z.literal('claimsSnapshot'), claims: z.array(Claim) }),
  z.object({ type: z.literal('setStatus'), claimId: z.string(), status: ClaimStatus }),
]);
export type ContentMessage = z.infer<typeof ContentMessage>;

File: services/backend/lale_backend/protocol.py
"""Wire protocol — Pydantic v2 mirror of packages/shared/src/protocol.ts.

Conventions:
  - Wire format is camelCase. We define snake_case fields with `alias` so JSON in/out is camelCase.
  - `populate_by_name=True` and `model_config = ConfigDict(populate_by_name=True, alias_generator=to_camel)`
    let us write Python in snake_case while the wire stays camelCase.
  - Discriminated unions match TS `z.discriminatedUnion("type", [...])`.

When you change a model here, change packages/shared/src/protocol.ts to match. Run /lale-protocol-check
to confirm parity before shipping.
"""

from __future__ import annotations

from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field
from pydantic.alias_generators import to_camel


class _Base(BaseModel):
    """All wire models inherit this — camelCase aliases, populate by name."""

    model_config = ConfigDict(
        alias_generator=to_camel,
        populate_by_name=True,
        extra="forbid",
    )


# ---------- Primitives ----------

ClaimType = Literal[
    "definition", "theorem", "lemma", "corollary", "proposition",
    "proof", "remark", "example", "claim", "conjecture",
    "exercise", "problem", "question",
]
ClaimStatus = Literal["unverified", "translating", "verifying", "verified", "failed", "sorry"]
RootCauseCategory = Literal[
    "unknownIdentifier",
    "typeMismatch",
    "tacticFailed",
    "timeout",
    "malformedJson",
    "other",
]
MathlibWorthinessVerdict = Literal[
    "mathlib_candidate",
    "project_local",
    "duplicate_likely",
    "needs_generalization",
    "not_worth_submitting",
    "unsure",
    "skipped",
]


# ---------- Claim ----------


class Claim(_Base):
    id: str
    type: ClaimType
    number: str | None = None
    label: str | None = None
    start_line: int = Field(ge=0)
    end_line: int = Field(ge=0)
    statement_latex: str
    proof_latex: str | None = None
    hash_latex: str
    hash_normalized: str
    status: ClaimStatus = "unverified"
    llm_dependency_ids: list[str] = Field(default_factory=list)


# ---------- /verify_paper ----------


class VerifyPaperRequest(_Base):
    request_id: UUID
    target_claim_id: str
    claims: list[Claim]
    lean_version: str
    mathlib_version: str


# ---------- SSE events ----------


class _PlanLevel(_Base):
    level: int = Field(ge=0)
    claim_ids: list[str]


class SseOrchestratorStarted(_Base):
    type: Literal["orchestratorStarted"]
    request_id: UUID
    target_claim_id: str
    plan: list[_PlanLevel]


class SseClaimStatus(_Base):
    type: Literal["claimStatus"]
    request_id: UUID
    claim_id: str
    status: ClaimStatus
    detail: str | None = None


class SseClaimVerified(_Base):
    type: Literal["claimVerified"]
    request_id: UUID
    claim_id: str
    elapsed_ms: int = Field(ge=0)
    cache_hit: bool
    lean_code: str | None = None


class SseMathlibWorthiness(_Base):
    type: Literal["mathlibWorthiness"]
    request_id: UUID
    claim_id: str
    verdict: MathlibWorthinessVerdict
    confidence: float = Field(ge=0.0, le=1.0)
    reason: str
    evidence: list[str] = Field(default_factory=list)
    suggested_location: str | None = None
    suggested_name: str | None = None
    review_notes: list[str] = Field(default_factory=list)


class SseClaimFailed(_Base):
    type: Literal["claimFailed"]
    request_id: UUID
    claim_id: str
    root_cause_category: RootCauseCategory
    deepest_failed_claim_id: str
    explanation: str
    lean_output: str | None = None
    lean_code: str | None = None


class SseOrchestratorFinished(_Base):
    type: Literal["orchestratorFinished"]
    request_id: UUID
    overall: Literal["verified", "failed", "partial"]


class SseClaimDependencies(_Base):
    type: Literal["claimDependencies"]
    request_id: UUID
    claim_id: str
    llm_dependency_ids: list[str]


SseEvent = Annotated[
    SseOrchestratorStarted
    | SseClaimStatus
    | SseClaimVerified
    | SseMathlibWorthiness
    | SseClaimFailed
    | SseOrchestratorFinished
    | SseClaimDependencies,
    Field(discriminator="type"),
]


# ---------- /status/{request_id} ----------


class StatusResponse(_Base):
    request_id: UUID
    state: Literal["running", "verified", "failed", "unknown"]
    events: list[SseEvent] = Field(default_factory=list)
